### PR TITLE
newlib-nano: fix include directory, 2nd attempt with pkg fix 

### LINKED ISSUE #9214: make: compiling is broken for any ARM platform on OS X.
#### Description

As per #9080, the newlib-nano directory is not included anymore in my local setup for
ARM platforms. Apparently the reason is that the commands used to add the directory
don't work properly on OS X. 
#### Steps to reproduce the issue

Compile anything on current master for any ARM platform on OS X.

#### Expected results

Compiles correctly.
#### Actual results

```console
arm-none-eabi-gcc \
	-DRIOT_FILE_RELATIVE=\"boards/samr21-xpro/board.c\" \
	-DRIOT_FILE_NOPATH=\"board.c\" \
	-Wall -Werror -Wextra -Wno-implicit-fallthrough -mno-thumb-interwork -mcpu=cortex-m0plus -mlittle-endian -mthumb -mfloat-abi=soft -march=armv6s-m -ffunction-sections -fdata-sections -fno-builtin -fshort-enums -ggdb -g3 -Os -std=c99 -fno-delete-null-pointer-checks -fdiagnostics-color -Wstrict-prototypes -Wold-style-definition -Werror=strict-prototypes -Werror=old-style-definition -fno-common -Wall -include '/Users/facosta/git/RIOT-OS/RIOT/examples/hello-world/bin/samr21-xpro/riotbuild/riotbuild.h'   -isystem  -I/Users/facosta/git/RIOT-OS/RIOT/core/include -I/Users/facosta/git/RIOT-OS/RIOT/drivers/include -I/Users/facosta/git/RIOT-OS/RIOT/sys/include -I/Users/facosta/git/RIOT-OS/RIOT/cpu/samd21/include -I/Users/facosta/git/RIOT-OS/RIOT/boards/samr21-xpro/include -I/Users/facosta/git/RIOT-OS/RIOT/cpu/sam0_common/include -I/Users/facosta/git/RIOT-OS/RIOT/cpu/cortexm_common/include -I/Users/facosta/git/RIOT-OS/RIOT/cpu/cortexm_common/include/vendor -I/Users/facosta/git/RIOT-OS/RIOT/sys/libc/include  -MD -MP -c -o /Users/facosta/git/RIOT-OS/RIOT/examples/hello-world/bin/samr21-xpro/board/board.o /Users/facosta/git/RIOT-OS/RIOT/boards/samr21-xpro/board.c
In file included from /Users/facosta/git/RIOT-OS/RIOT/boards/samr21-xpro/include/board.h:26:0,
                 from /Users/facosta/git/RIOT-OS/RIOT/boards/samr21-xpro/board.c:23:
/Users/facosta/git/RIOT-OS/RIOT/cpu/cortexm_common/include/cpu.h:35:10: fatal error: irq.h: No such file or directory
 #include "irq.h"
          ^~~~~~~
compilation terminated.
```
#### Versions

```
Installed toolchain versions
----------------------------
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
          native gcc: Apple LLVM version 8.0.0 (clang-800.0.42.1)
          msp430-gcc: msp430-gcc (GCC) 4.7.0 20120322 (mspgcc dev 20120911)
             avr-gcc: avr-gcc (GCC) 7.3.0
   arm-none-eabi-gcc: arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 7-2017-q4-major) 7.2.1 20170904 (release) [ARM/embedded-7-branch revision 255204]
    mips-mti-elf-gcc: missing/error
               clang: Apple LLVM version 8.0.0 (clang-800.0.42.1)
arm-none-eabi-newlib: "2.5.0"
 mips-mti-elf-newlib: missing/error
            avr-libc: "2.0.0" ("20150208")
            cppcheck: missing
          coccinelle: missing
                 git: git version 2.17.0
```

### PR DESCRIPTION
**Pull request by smlng**
### Contribution description

based on #9216, but including fixes for cmake based external packages

### Issues/PRs references

- includes #9216
- fixes #9214

### CODE REVIEW

**Comment by smlng** [comment]
btw. includes all commits originally by @toonst

**Comment by smlng** [comment]
there seems to be problem with the MIPS toolchain and its `_Atomic(t)` macro in conjunction with the newlib stuff, somehow all the typedefs  for `int_least8_t` (and others) are not found anymore 😩

**Comment by smlng** [comment]
However this is currently only a problem with `tests/rmutex` which is why I blacklisted the 3 MIPS boards in the last commit.

File: pkg/ccn-lite/Makefile [hunk 1]
```diff
@@ -5,7 +5,7 @@ PKG_LICENSE=ISC
 
 .PHONY: all
 
-export RIOT_CFLAGS = $(CFLAGS) $(INCLUDES)
+RIOT_CFLAGS := $(INCLUDES)
```
Comment (josephnoir) [inline pkg/ccn-lite/Makefile]: This should be probably be renamed to `RIOT_INCLUDES` to reflect its contents.

Comment (smlng) [reply pkg/ccn-lite/Makefile]: mhm, those are still `CFLAGS` passed to the C compiler

Comment (josephnoir) [reply pkg/ccn-lite/Makefile]: hmh, I can accept that. Still need someone else to review the newlib part, any suggestions who?

**Comment by tcschmidt** [comment]
@smlng can you identify a proper additional reviewer?

**Comment by cgundogan** [comment]
>there seems to be problem with the MIPS toolchain and its _Atomic(t) macro in conjunction with the newlib stuff, somehow all the typedefs for int_least8_t (and others) are not found anymore

>However this is currently only a problem with tests/rmutex which is why I blacklisted the 3 MIPS boards in the last commit.

@smlng but essentially this breaks mips in conjunction with `int_least8`. I don't think blacklisting these boards is the correct way, because user applications outside of the RIOT source directory – and thus hidden to us – may break. It's probably just due to some reordering of the includes..

@neiljay are you familiar with the nano.specs newlib + MIPS or did you encounter a similar issue?

File: makefiles/libc/newlib.mk [hunk 2]
```diff
@@ -45,29 +54,32 @@ NEWLIB_INCLUDE_PATTERNS ?= \
...
-  # newlib-nano overrides newlib.h and its include dir should therefore go before
-  # the regular system include dirs.
-  INCLUDES := -isystem $(NEWLIB_NANO_INCLUDE_DIR) $(INCLUDES)
+  ifeq (1,$(USE_NEWLIB_NANO))
+    # newlib-nano include directory is called either newlib-nano or nano. Use the one we find first.
+    NEWLIB_NANO_INCLUDE_DIR ?= $(firstword $(wildcard $(addprefix $(NEWLIB_INCLUDE_DIR)/, newlib-nano nano)))
+    ifneq ($(NEWLIB_NANO_INCLUDE_DIR),)
+      # newlib-nano overrides newlib.h and its include dir should therefore go before
+      # the regular system include dirs, but after the other newlib includes.
+      NEWLIB_INCLUDES += -isystem $(NEWLIB_NANO_INCLUDE_DIR)
```
Comment (gebart) [inline makefiles/libc/newlib.mk]: Why is the nano include dir after the regular newlib dir? The comment does not make sense

File: makefiles/libc/newlib.mk [hunk 3]
```diff
@@ -1,7 +1,15 @@
+ifeq ($(BOARD), native)
+  $(error Newlib is not supported on native target)
+endif
+
 ifneq (,$(filter newlib_nano,$(USEMODULE)))
+  USE_NEWLIB_NANO = 1
   # Test if nano.specs is available
   ifeq ($(shell $(LINK) -specs=nano.specs -E - 2>/dev/null >/dev/null </dev/null ; echo $$?),0)
-    USE_NEWLIB_NANO = 1
+    # Clang/llvm has no support for specs files
```
Comment (gebart) [inline makefiles/libc/newlib.mk]: Shouldn't the test above already filter out any compiler which errors on -specs?

File: makefiles/libc/newlib.mk [hunk 4]
```diff
@@ -45,29 +54,32 @@ NEWLIB_INCLUDE_PATTERNS ?= \
...
 # directories as system include directories, which means they will not be
 # searched until after all the project specific include directories (-I/path)
-NEWLIB_INCLUDE_DIR ?= $(firstword $(wildcard $(NEWLIB_INCLUDE_PATTERNS)))
+NEWLIB_INCLUDE_FILE = $(firstword $(wildcard $(addsuffix /newlib.h,$(NEWLIB_INCLUDE_PATTERNS))))
 
 # If nothing was found we will try to fall back to searching for a cross-gcc in
 # the current PATH and use a relative path for the includes
-ifeq (,$(NEWLIB_INCLUDE_DIR))
+ifeq (,$(NEWLIB_INCLUDE_FILE))
+  $(warning Using relative include dir)
```
Comment (gebart) [inline makefiles/libc/newlib.mk]: Suggestion: change the message to "newlib.h not found, guessing newlib include path from gcc path"

**Comment by neiljay** [comment]
@cgundogan Hi, No I have no experience with newlib-nano I checked with our compiler team and they said they have not tried nanospecs either, with the MIPS toolchain we do supply a SmallClib (-mclib=small) and TinyClib (-mclib=tiny) which are newlib derived if you need something smaller, but these are not the same as newlib-nano though.

File: makefiles/libc/newlib.mk (hunk 3 above)
Comment (toonst) [reply makefiles/libc/newlib.mk]: Even though it has no support for it, Clang does not fail when passed the  -specs option. Adding -Werror makes sure Clang fails when the -specs flag is passed.

**Comment by smlng** [comment]
closing in favour of original PR #9216

**Pull request closed without merging**